### PR TITLE
Flask2.0

### PR DESCRIPTION
**Pull request by maringuu**
~Depends on #618 but it could easily be rebased on master~

### CODE REVIEW

File: src/web_interface/frontend_main.py [hunk 1]
```diff
@@ -27,10 +27,11 @@ def __init__(self, config=None):
         logging.info("Web front end online")
 
     def _setup_app(self):
+        # TODO replace __name__ see Flask constructor docs
         self.app = Flask(__name__)
         self.app.config.from_object(__name__)
 
-        Flask.secret_key = os.urandom(24)
```
Comment (maringuu) [inline src/web_interface/frontend_main.py:33]: I wonder how this worked, since it does not actually set the `secret_key` in `app.config`

**Comment by dorpvom** [comment]
I've checked this and it doesn't work. Scenario:
1. I create a user and role according to the documentation.
2. I use the current master with activated authentication.
3. I can successfully log into FACT using the create credentials.
4. I switch to this branch and run the installation (after manually removing flask_security)
5. I can't log into FACT using the credentials or even credentials created after switching the branch.

**Comment by dorpvom** [comment]
We have to look for reasons why this behavior occurs. Maybe flask-security-too has a differend handle on the database.

**Comment by maringuu** [comment]
I added alembic to handle database migrations. It might be a bit overkill but it seems to be standart in the flask ecosystem.

To upgrade the database run `alembic upgrade head`.
For downgrading use `alembic downgrade base`.

For me the problem is fixed.

@dorpvom Can you confirm?

**Comment by dorpvom** [comment]
> 
> 
> I added alembic to handle database migrations. It might be a bit overkill but it seems to be standart in the flask ecosystem.
> 
> To upgrade the database run `alembic upgrade head`.
> For downgrading use `alembic downgrade base`.
> 
> For me the problem is fixed.
> 
> @dorpvom Can you confirm?

We'll discuss this tomorrow

**Comment by maringuu** [comment]
Note that the tests fail in the CI the packages are not upgraded in the CI.

File: src/migrate_database.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    users = [row[0] for row in cur.execute('SELECT id FROM user')]
+    for uid in users:
+        cur.execute('UPDATE user SET fs_uniquifier = ? WHERE user.id = ?', (uuid.uuid4().hex, uid))
+
+    # Due to limitiations in SQLite we have to create a temporary table
+    cur.execute('''
+        CREATE TABLE "user_tmp" (
+            "id"			INTEGER NOT NULL,
+            "api_key"		VARCHAR(255),
+            "email"			VARCHAR(255),
```
Comment (jstucke) [inline src/migrate_database.py:22]: shouldn't these be `UNIQUE`?

File: src/test/unit/web_interface/security/test_user_role_db_interface.py [hunk 3]
```diff
@@ -14,6 +14,7 @@ def setUp(self):
         self.test_app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///'
         self.test_app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
         self.test_app.config['SECURITY_PASSWORD_SALT'] = 'salt123'
+        self.test_app.config['SECRET_KEY'] = 'key123'
```
Comment (jstucke) [inline src/test/unit/web_interface/security/test_user_role_db_interface.py:17]: > test/unit/web_interface/security/test_user_role_db_interface.py:7: in <module>
    from web_interface.security.authentication import create_user_interface
E   ImportError: cannot import name 'create_user_interface' from 'web_interface.security.authentication'

File: src/web_interface/security/authentication.py [hunk 4]
```diff
@@ -1,27 +1,35 @@
...
+    # flask-security we change the validation mapper of 'email'.
+    app.config['SECURITY_USER_IDENTITY_ATTRIBUTES'] = [{'email': {"mapper": uia_username_mapper, "case_insensitive": True}}]
 
-    user_interface = create_user_interface(db)
-    security = Security(app, user_interface)
 
-    _add_apikey_handler(security, user_interface)
-    return db, user_interface
+def add_flask_security_to_app(app):
+    # db.Table etc cause errors but are correct
```
Comment (jstucke) [inline src/web_interface/security/authentication.py:28]: I did not observe any errors. Should I see errors?

**Review by jstucke — COMMENTED** [review]
Everything seems to work. It might be a problem to directly upgrade, because `flask-security` might still be installed. Maybe we make sure that it is uninstalled during the installation?

File: src/migrate_database.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            "password"		VARCHAR(255),
+            "active"		BOOLEAN,
+            "confirmed_at"	DATETIME,
+            "fs_uniquifier"	VARCHAR(64)	NOT NULL,
+            CHECK(active IN (0,1)),
+            PRIMARY KEY("id")
+        );''')
+    cur.execute('INSERT INTO "user_tmp" SELECT * FROM "user" WHERE true')
+    cur.execute('DROP TABLE "user"')
+    cur.execute('ALTER TABLE "user_tmp" RENAME TO "user"')
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/migrate_database.py:34]: The ALTER TABLE statement is very limeted in SQLite. I added an explaining comment above.

File: src/web_interface/security/authentication.py (hunk 4 above)
Comment (maringuu) [reply src/web_interface/security/authentication.py:28]: For me pylint complains that `Table` is not a member of `db`.
Pylint is wrong so I disabled the warnings and added this comment to explain why I disabled them.

File: src/web_interface/frontend_main.py [hunk 6]
```diff
@@ -27,11 +27,13 @@ def __init__(self, config=None):
         logging.info("Web front end online")
 
     def _setup_app(self):
+        # FIXME replace __name__ see Flask constructor docs
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/web_interface/frontend_main.py]: I didn't know about `flask_app_wrapper.py`. I think you are right. I removed the comment.

**Comment by maringuu** [comment]
I don't think Lintly is correct here.
When conforming to the standart (I used autopep8) it becomes less readable imo.

File: src/test/unit/test_manage_users.py [hunk 7]
```diff
@@ -51,9 +54,24 @@ def test_integration_try_actions(monkeypatch, action_and_inputs):
...
+        {'data_storage': {
+            # We want an in memory database for testing
+            'user_database': 'sqlite://',
+            'password_salt': "salt"
+            },
+        'ExpertSettings': {
+            'authentication': 'true'
+            },
+        }
+    )
```
Comment (jstucke) [inline src/test/unit/test_manage_users.py:70]: indentation fix:
```suggestion
    parser.read_dict({
        'data_storage': {
            # We want an in memory database for testing
            'user_database': 'sqlite://',
            'password_salt': "salt"
        },
        'ExpertSettings': {
            'authentication': 'true'
        },
    })
```

**Review by jstucke — APPROVED** [review]

**Pull request merged**